daemons/dss-sync-sfn/app.py

```diff
@@ -0,0 +1,306 @@
+            "Next": "ComposeParts"
+        },
+        "ComposeParts": {
+            "Type": "Task",
+            "Resource": None,  # This will be set by Domovoi to the Lambda ARN
+            "End": True,
+            "Retry": retry_config
+        },
+        "Quit": {
+            "Type": "Pass",
+            "End": True
+        }
+    }
+}
+
+sfn_thread = {
+    "StartAt": "Worker{t}",
+    "States": {
+        "Worker{t}": {
+            "Type": "Task",
+            "Resource": None,  # This will be set by Domovoi to the Lambda ARN
+            "Next": "Branch{t}",
+            "Retry": retry_config
+        },
+        "Branch{t}": {
+            "Type": "Choice",
+            "Choices": [{
+                "Variable": "$.finished",
+                "BooleanEquals": True,
+                "Next": "EndThread{t}"
+            }],
+            "Default": "Worker{t}"
+        },
+        "EndThread{t}": {
+            "Type": "Pass",
+            "End": True
+        }
+    }
+}
+
+@app.step_function_task(state_name="DispatchSync", state_machine_definition=sync_sfn_num_threads)
+def dispatch_sync(event, context):
+    """
+    Processes the storage event notification and orchestrates the rest of the copying:
+    - If the notification is for a file blob:
+        - If the blob is under the one-shot threshold, immediately copies it and exits the state machine.
+        - Otherwise, configures the state machine to run the threadpool (copy blob parts) and closer (compose the copy).
+    - Otherwise, check if all referenced entities are already copied to the destination:
+        - For a file manifest, check that the blob is there.
+        - For a bundle manifest, check that file manifests for all files in the bundle are there.
+        - For a collection manifest, check that all collection contents are there.
+        If the checks fail, cause the state machine to sleep for 8 seconds, then try again.
+        If the checks succeed, do a one-shot copy of the manifest.
+    """
+    if event["source_key"].startswith("blobs"):
+        if int(event["source_obj_metadata"]["size"]) > AWS_MIN_CHUNK_SIZE:
+            new_event = dict(event, sleep=False, do_oneshot_copy=False, do_multipart_copy=True, done=False)
+            if Replica[event["dest_replica"]] == Replica.aws:
+                mpu_id = initiate_multipart_upload(source_replica=Replica[event["source_replica"]],
+                                                   dest_replica=Replica[event["dest_replica"]],
+                                                   source_key=event["source_key"])
+                new_event.update(mpu_id=mpu_id)
+            return new_event
+        else:
+            return dict(event, sleep=False, do_oneshot_copy=True, do_multipart_copy=False, done=False)
+    return dict(event, sleep=False, do_oneshot_copy=False, do_multipart_copy=False, done=False)
+
+def copy_parts(event, context):
```

Comment:
There is logical and conceptual ovrerlap here with `dss.stepfunctions.s3copyclient` and `dss.stepfunctions.gscopyclient`. Multipart copy is a fundamental operation that I feel should be defined in one place.

This is just echoing @Bento007's comments, which have already received responses.

Reply:
The copy clients in dss.stepfunctions copy data by reference, while the sync daemon copies parts by value. Because of this fundamental design requirement, the codebases were engineered differently. While there is otherwise some conceptual overlap, the sync daemon uses low level streaming APIs to copy the data, and the organizational differences in the codebases are unfortunately quite hard to reconcile.

As I mentioned in my responses to @Bento007, I don't believe refactoring this codebase to conform to other portions of the DSS code should be the focus of this PR. This PR does the work necessary to port sync daemon to step functions. We can separately address these functional commonality concerns between `dss.stepfunctions.*copyclient`, `blobstore` and this codebase later.